Fix EZP-24753: non-admin load user on login to PlatformUI

joaoinacio opened the pull request:
JIRA: https://jira.ez.no/browse/EZP-24753

andrerom commented on eZ/Publish/Core/REST/Server/Controller/User.php:
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
+        $mainLocationId = $userContentInfo->mainLocationId;
+
+        // special case for loading current user, which does not have permissions to load its location.
+        if ($user == $this->repository->getCurrentUser()) {
```
this condition will always be true, there is always a user.

joaoinacio replied:
@andrerom : I'm not sure I understand.

$user will be the user which is requested for the REST request;
currentUser should be the current logged in user, or anonymous.

With the above, if user with ID 50 tries to load f.e. user with ID 10, the condition will be false?

andrerom replied:
Ah, I didn't see the `==`, thought you where doing a `=`, please make the sentance more clear by doing:
`if ($user->id === $this->repository->getCurrentUser()->id) {`

lolautruche replied:
+1 with @andrerom

lolautruche commented on eZ/Publish/Core/REST/Server/Controller/User.php:
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
+                    return $locationService->loadLocation($mainLocationId);
+                }
+            );
+            $relations = array();
```
Why do you leave `$relations` empty here?

joaoinacio replied:
Not really certain how relations should be handled.

If loading the own user's location shouldn't pose any problem, using sudo to load relations could expose restricted content, hence a potential security issue?

lolautruche replied:
Right, +1 for leaving it empty.

andrerom replied:
please add inline comment on this

andrerom replied:
/cc @bdunogier on this ;)

lolautruche commented on eZ/Publish/Core/REST/Server/Controller/User.php:
```diff
@@ -179,16 +179,31 @@ public function loadUser($userId)
+
+        // special case for loading current user, which does not have permissions to load its location.
+        if ($user->id == $this->repository->getCurrentUser()->id) {
+            $locationService = $this->locationService;
```
This is not needed any more as we require PHP 5.4 as of eZ 5.4, unless we want to backport this to 5.3.

andrerom replied:
Are you implying binding $this to the closure?

lolautruche replied:
Yes, `$this` is correctly dereferenced in closures as of PHP 5.4

lolautruche commented:
+1 Besides my last nitpick.
Travis failure doesn't seem to be related.

Any idea @andrerom @vidarl ?

Ping @bdunogier @dpobel @yannickroger @glye

lolautruche commented:
And of course ping @pspanja

pspanja commented:
+1 with last remark on closure resolved.

joaoinacio commented:
Updated according to feedback, also modified the flow a bit (sorry!) to only try sudo if the first load fails.
This allows f.e. relations to be loaded for admin user, etc - only if it fails is sudo attemped.
(AFAIK there should be no perf considerations on login).

If needed the change can be reverted, sorry again and thanks for the review

andrerom commented:
@joaoinacio please don't change things like that after review, that makes the review worthless.
It looks slightly more of a hack now, but as it is hack anyway, so +1 BUT:

please change your comment like this, and no other line:
```diff
- // special case: allow current logged in user to load location for its own User content
+ // TODO: Hack for special case to allow current logged in user to load him/here self (but not relations)
```

lolautruche commented:
+1

joaoinacio commented:
Done.

bdunogier commented:
Hack indeed, but let's say +1.

It would be nice to cover in BDD that non-admin editors should be able to login to PlatformUI, even if it is as a follow-up.

andrerom commented:
> It would be nice to cover in BDD that non-admin editors should be able to login to PlatformUI, even if it is as a follow-up.

very good point, @miguelcleverti could that be the QA part on this story maybe? *(instead of manual tests)*

andrerom commented:
Merging, thanks @joaoinacio !
